app/javascript/controllers/letters_calls_form_controller.js

```diff
@@ -0,0 +1,96 @@
+// Get the value of an element attribute as a number or a default value
+function getAttributeAsNumber(element, attributeName, defaultValue = 0) {
+	return parseFloat(element.getAttribute(attributeName)) || defaultValue;
+}
+
+// Calculate the total based on the values of letters and phone_calls inputs
+function calculateTotal(lettersValue, phoneCallsValue, unitPriceForLetters, unitPriceForPhoneCalls) {
+	const total = lettersValue * unitPriceForLetters + phoneCallsValue * unitPriceForPhoneCalls;
+	return parseFloat(total.toFixed(2)); // Format total to two decimal places
+}
+
+// Calculate the total with uplift
+function calculateTotalWithUplift(currentTotal, upliftPercentage) {
+	const upliftToAdd = currentTotal * (upliftPercentage / 100);
+	const totalWithUplift = (currentTotal + upliftToAdd);
+	return totalWithUplift;
+}
+
+// Update the DOM element with the provided total
+function updateDomElementTotal(totalAmount, element) {
+	totalAmount = parseFloat(totalAmount).toFixed(2);
+	element.innerHTML = '£' + totalAmount;
+}
+
+// Check if all necessary field values exist
+function checkIfAllFieldValueExists(lettersValue, phoneCallsValue) {
+	return lettersValue && phoneCallsValue;
+}
+
+// Handle change events on the form
+export function handleFormChange(event, form, totalElement, lettersElement, phoneCallsElement, upliftCheckboxElement, upliftPercentageElement) {
+	const lettersValue = parseFloat(lettersElement.value) || 0;
+	const phoneCallsValue = parseFloat(phoneCallsElement.value) || 0;
+	const unitPriceForLetters = getAttributeAsNumber(lettersElement, 'data-rate-letters');
+	const unitPriceForPhoneCalls = getAttributeAsNumber(phoneCallsElement, 'data-rate-phone-calls');
+	const upliftPercentage = parseFloat(upliftPercentageElement.value) || 0;
+
+	if (checkIfAllFieldValueExists(lettersValue, phoneCallsValue)) {
+		const currentTotal = calculateTotal(lettersValue, phoneCallsValue, unitPriceForLetters, unitPriceForPhoneCalls);
+		// Triggered when letters or phone calls value changes, or when uplift percentage changes
+		if ([lettersElement, phoneCallsElement, upliftPercentageElement].includes(event.target)) {
+			// Update total
+			const totalWithUplift = calculateTotalWithUplift(currentTotal, upliftPercentage);
+			updateDomElementTotal(totalWithUplift, totalElement);
+		}
+		if (event.target === upliftCheckboxElement && !upliftCheckboxElement.checked) {
+			// Clear uplift percentage value
+			upliftPercentageElement.value = '';
+			// Update total
+			const totalWithUplift = calculateTotalWithUplift(currentTotal, 0);
+			updateDomElementTotal(totalWithUplift, totalElement);
+		}
+	} else {
+		// Called when letters or phone calls value is empty
+		updateDomElementTotal(0, totalElement)
+	}
+}
```

Comment:
I found the multiple paths that could lead to an update to the HTML (it is called in 3 spots) slightly hard to follow. Can I suggest that refactoring to something like the below might make it easier to others to follow:

```javascript
// Handle change events on the form
export function handleFormChange(event, form, totalElement, lettersElement, phoneCallsElement, upliftCheckboxElement, upliftPercentageElement) {
	// check if we care about this event
	if ([lettersElement, phoneCallsElement, upliftPercentageElement, upliftCheckboxElement].includes(event.target)) {

		// reset uplift percentage if required
		if (event.target === upliftCheckboxElement && !upliftCheckboxElement.checked) {
			// Clear uplift percentage value
			upliftPercentageElement.value = '';
		}

		// load the values so we can check them
		const lettersValue = parseFloat(lettersElement.value) || 0;
		const phoneCallsValue = parseFloat(phoneCallsElement.value) || 0;
		let totalWithUplift;

		// calculate the total value
		if (checkIfAllFieldValueExists(lettersValue, phoneCallsValue)) {
			// load the remaining values as we are now going to use them
			const unitPriceForLetters = getAttributeAsNumber(lettersElement, 'data-rate-letters');
			const unitPriceForPhoneCalls = getAttributeAsNumber(phoneCallsElement, 'data-rate-phone-calls');
			const upliftPercentage = parseFloat(upliftPercentageElement.value) || 0;

			// update the values
			totalWithUplift = calculateTotalWithUplift(currentTotal, upliftPercentage);
		} else {
			totalWithUplift = 0			
		}

		//update screen with calculated value
		updateDomElementTotal(totalWithUplift, totalElement);
	}	
}
```